미션 제출합니다

iimpala opened the pull request:
아직까지는 개선할 점이 많은 코드라고 느껴집니다.

객체지향적인 설계와 리소스관리, 예외처리가 부족하다고 느끼는데 이부분에 집중해서 리뷰 남겨주시면 감사하겠습니다.
추가로 코드를 작성할 때 자주 쓰이는 관례? 같은 것들도 알려주시면 감사하겠습니다!

pobiconan committed:
initial commit

javajigi committed:
docs: 미션 제목 변경

javajigi committed:
add nextstep utils, test code, update library version

javajigi committed:
refactor test code

javajigi committed:
refactor names test data

javajigi committed:
refactor winner message

javajigi committed:
refactor desc to main method

javajigi committed:
update the latest version

javajigi committed:
remove Car class

javajigi committed:
fix(racingcar): correct the test expected text

chanhalee committed:
docs(README) 요구사항 정리

페어와 요구사항을 정리함

chanhalee committed:
build(gradle) Java 버전 및 gradle 버전 수정

Java 17, gradle 7.2로 업데이트

chanhalee committed:
feat(Car) Car 전진 요구사항 테스트 작성 및 구현

자동차가 전진하면 위치가 1 증가한다

chanhalee committed:
docs(README) 요구사항 업데이트

요구사항 구체화

chanhalee committed:
feat(NumberGenerator) 랜덤 숫자 생성기 테스트 및 구현

0에서 9 사이의 정수를 생성하는 숫자 생성기 테스트 제작과 구현

chanhalee committed:
feat(ProceedLogic) 전진 여부를 판정하는 로직 테스트 작성 및 구현

전진 여부를 판정하는 로직 테스트 제작과 구현

chanhalee committed:
feat(RacingGame) 자동차들 중 가장 많이 이동한 자동차가 우승자가 된다.

자동차들 중 가장 많이 이동한 자동차가 우승자가 된다.
- 복수의 우승자가 발생할 수 있다.

chanhalee committed:
docs(README) 진행 현황 업데이트

진행 현황 업데이트

chanhalee committed:
refactor(Cars) 클래스 이름 변경

클래스 역활 일부 수정에 의한 명칭 변경
RacingGame -> Cars

chanhalee committed:
feat(UI) CLI 인터페이스 구현

사용자와 상호작용할 CLI 인터페이스 구현
- CarState dto 클래스 생성

chanhalee committed:
feat(UI) 에러 메시지 출력 ui작성

각 상황별 에러메시지를 출력하는 ui구현

chanhalee committed:
feat(controller) ui와 도메인을 연결하는 컨트롤러 로직 작성

사용자 입력 검증 및 레이싱 게임 실행 로직 구현

chanhalee committed:
docs(README) 프로젝트 진행 현황 업데이트

프로젝트 진행 현황 업데이트

chanhalee committed:
refactor(RacingGame) validation 로직 리팩토링

validation 로직 리팩토링

chanhalee committed:
docs(README) 문자열 덧셈 계산기 TODO 작성

문자열 덧셈 계산기 TODO 작성

chanhalee committed:
feat(StringParser) 문자열 파서 구현

입력된 문자열을 해석해서 정수형 리스트를 반환하는 파서 구현

chanhalee committed:
feat(StringCalculator) 문자열 덧셈 계산기 구현

입력된 문자열을 해석해서 덧셈 연산 결과를 반환하는 계산기 구현

chanhalee committed:
refactor(StringParsor) 파서 리팩토링

코드 컨벤션을 준수하도록 리팩토링

chanhalee committed:
fix: git repository different conflict

seokhong commented on README.md:
```diff
@@ -1,9 +1,92 @@
+
+## Todo
+
+## Done
```
기능 목록을 작성할 때 마크다운의 체크리스트 문법을 활용해보면 어떨까요?
Todo 에서 Done 으로 옮기는 방법과 비슷하지만, 조금 더 쉽게 체크할 수 있고, 커밋 단위로 변경도 보기 쉽다는 장점이 있습니다. :)

iimpala replied:
마크다운에 그런 기능이 있었군요! 리뷰 감사합니다:)

[471ed6d](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/471ed6de4b0a1bc117a15cf4c3d9e79235c8d265)

seokhong commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,9 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+<<<<<<< HEAD
```
git merge 과정에서 해결되지 않은 충돌이 있는 것으로 보이네요!

iimpala replied:
반영했습니다!

[5442f9a](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/5442f9ad2b4cab391dade9c6df8567359e454b40)

seokhong commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	@Override
+	public int compareTo(Car other) {
```
`Car#compareTo`의 구현을 모르는 개발자나 외부 클래스에서 이 메서드를 사용할 때, 자동차를 비교하면 위치 값을 비교한다고 생각하기 어렵지 않을까요? 명시적으로 `comparePosition` 과 같은 메서드를 사용해보는 건 어떨까요? `Comparable`을 구현하신 이유가 있는지 궁금합니다. :)

iimpala replied:
우승자를 찾는 메소드에서 정렬을 통해 우승자를 확인할 수 있도록 구현하기 위해 Comparable을 구현했습니다.

그런데 리뷰를 보니 의미가 compareTo 만으로는 불분명할 수 있을 것 같네요. 리뷰 감사합니다:)

[480fd1d](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/480fd1d1cca72285b6e12bfda0ec56217209670b)

seokhong commented on src/main/java/racingcar/ErrorType.java:
```diff
@@ -0,0 +1,19 @@
+package racingcar;
+
+public enum ErrorType {
```
에러 메시지를 사용하는 곳에서 상수로 정의하지 않고, 별도의 타입으로 분리하신 이유가 있을까요?

iimpala replied:
프로그램 전체에 대한 에러 메시지를 한 곳에서 관리하기 위해서 글로벌로 ErrorType을 선언했습니다.
관리의 측면에서는 이게 더 용이할 것 같아서 이런식으로 작성했는데 혹시 에러메시지를 사용하는 곳에서 상수로 정의하면 어떤 이점이 있을까요??

seokhong replied:
저는 상수를 한 곳으로 모았을 때 부작용이 더 크게 느껴져서, 가능하면 사용하는 곳에서 관리하는 편입니다.
1. 상수를 선언하는 곳과 사용하는 곳이 분리되어 있으면, 사용하는 곳이 사라졌을 때 선언하는 곳도 함께 관리를 해야하는데, 관리가 잘 안 됐던 경험이 있습니다.
2. 예외 메시지도 도메인의 요구사항에 맞춰 정의되는데, 상수를 별도의 클래스로 분리하면 요구사항이 달라졌을 때 변경해야 할 지점이 두 곳으로 늘어납니다.
3. 상수의 개수(예외 메시지의 개수)가 많아지면 상수 정의 클래스의 길이가 길어지고, 이렇게 메시지만 정의된 클래스의 코드가 길어질 때 관리가 잘 안 됐던 경험이 있습니다.

seokhong commented on src/main/java/racingcar/NumberGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar;
+
+public interface NumberGenerator {
```
인터페이스를 분리하고 테스트에서 값 생성을 제어하신 점 좋습니다. 👍

seokhong commented on src/main/java/racingcar/ProceedLogic.java:
```diff
@@ -0,0 +1,12 @@
+	}
+
+	public boolean askProceed() {
+		return numberGenerator.generate() > 3;
```
3이라는 값도 도메인에서 의미를 가지는 값으로 보이니, 상수로 변경하고 이름을 부여해보면 어떨까요?

iimpala replied:
반영했습니다. 리뷰 감사합니다:)

[076d0e0](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/076d0e043d9d348dfb452c71b946b1ad498ec4e3)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+		this.logic = new ProceedLogic(numberGenerator);
+	}
+
+	public void play() {
```
메서드의 배치는 어떤 순서로 하시나요?

iimpala replied:
되도록이면 public 메소드를 맨 위에 모아두고 아래에 private 메소드를 모으는 식으로 정리하려하고 있습니다.

seokhong replied:
말씀하신 방식으로 정리했을 때의 장점은 어떤 것이 있었나요? :)
저는 public 메서드 배치 후, 해당 메서드에서 사용하는 private 메서드를 순차적으로 배치하는데요. 이렇게 하면 코드를 읽을 때 위에서부터 순서대로 읽고 이해하기 용이하다는 장점이 있습니다. 관련해서 클린 코드의 메서드 배치 순서에 대한 내용을 참고해보셔도 좋겠습니다.

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+
+	private void initCars() {
+		boolean validationPass = false;
+		List<String> carNameList = null;
```
변수의 이름에 구체적인 컬렉션의 타입(e.g.,`Set`,`List`)은 사용하시지 않는 것을 추천합니다.  
리팩터링 과정에서 타입이 변경되기도 하는데, 이때 이름도 함께 변경해줘야 해서 비용이 추가로 발생합니다.  
`carNames`와 같은 이름을 사용해보는 건 어떨까요?

iimpala replied:
컬렉션이 바뀔수 있다는 생각은 해보지 않았는데 그럴수도 있겠네요! 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+		return notEmptyRound(round) && notTooLongRound(round) && isNumericalRound(round);
+	}
+
+	private boolean validateCarNames(List<String> carNameList) {
```
`validate`, `check` 등으로 시작하는 메서드는 대개 밸리데이션을 수행하는 메서드의 prefix 로 많이 사용합니다.
밸리데이션을 수행하는 메서드는 값을 반환하지 않고, 문제 상황인 경우 예외를 발생시키는 형태로 많이 구현합니다.
boolean 값을 반환하는 메서드의 경우 `is`, `has` prefix 를 주로 사용합니다. 자동차의 이름이 유효한지 여부를 반환하는 메서드이니, `isValidCarNames`와 같은 이름을 사용해보는 건 어떨까요?

iimpala replied:
아하 그렇군요! 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+		return notEmptyCarName(carNameList) && notTooLongCarName(carNameList) && isUniqueCarName(carNameList);
+	}
+
+	private boolean notEmptyRound(String round) {
```
`not` 을 붙이는 경우, 정상 케이스를 반대로 한 번 더 생각해야 해서 코드를 이해하는데 에너지를 조금 더 사용해야 합니다.
`isValidRound` 로 변경하고, 정상인 경우 true 를 반환하는 형태로 구현해보는 건 어떨까요?

iimpala replied:
리뷰를 받고 다시보니 확실히 이해가 어렵긴 하네요. 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	private boolean notTooLongRound(String round) {
+		if (round.length() > 5) {
```
자동차 경주의 라운드가 5를 초과할 수 없다는 조건은 어디에 있을까요? :)

iimpala replied:
라운드의 상한선에 대한 별도의 언급이 없어서 임의로 추가한 내용입니다.

사용자의 입력 문자열을 Integer.parseInt()를 통해 정수형으로 변환해야 하는데
Integer의 최댓값을 넘어가는 입력이 들어오는 경우에 대한 처리를 어떤 방식으로 해야할지 감이 잘 잡히지 않아서 임의의 상한선을 두었습니다.

혹시 이런 경우는 어떤식으로 처리하는게 좋을까요??

seokhong replied:
말씀하신 것처럼 상한선이 필요한 케이스라면, 그 경우를 정의하고 밸리데이션을 추가하는 것 좋습니다. 👍 
다만 Integer 의 범위로 인한 상한선이라면, Long 이나 다른 타입을 이용하는 것이 더 바람직할 수도 있다는 생각이 듭니다. 타입의 제한으로 인해 도메인이 제한되는 것이 바람직하다고 생각하지는 않습니다.
오히려 자동차 게임은 최대 30라운드까지만 제공합니다와 같이 도메인에서 적절하다고 생각하는 횟수를 정하고 제한하는 것이 더 좋다고 생각합니다. 그리고, 이와 비슷한 이유로 5라운드로 제한하신 것으로 이해했습니다! :)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	private boolean notEmptyCarName(List<String> carNameList) {
+		if (carNameList.isEmpty() || !carNameList.stream().filter(String::isEmpty).toList().isEmpty()) {
```
`!carNameList.stream().filter(String::isEmpty).toList().isEmpty()`로 한 줄에 작성할 수 있는 코드이지만, 한 눈에 읽고 이해하기는 어려운 것 같아요.
코드의 줄 수가 늘어나더라도, 조금 더 가독성이 좋도록 지역 변수나 메서드 등으로 분리해보는 건 어떨까요?

iimpala replied:
지역변수로 추출하여 리팩토링했습니다. 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	private boolean notTooLongCarName(List<String> carNameList) {
+		if (!carNameList.stream().filter(name -> name.length() > 5).toList().isEmpty()) {
```
5라는 값도 상수로 분리할 수 있을 것으로 보이네요!

iimpala replied:
반영했습니다. 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	public Car(String name, int position) {
+		this.name = name;
```
기능 요구사항 중 자동차 이름에 대한 조건이 있네요.
도메인에서도 이름의 길이를 검증해보면 어떨까요?

- 자동차 이름은 쉼표(,)를 기준으로 구분하며 이름은 5자 이하만 가능하다.

iimpala replied:
사용자로부터 입력을 받을때 검증로직이 들어있긴 한데 도메인 레이어에서도 추가로 해주는게 좋겠군요! 리뷰 감사합니다:)

[d0fb520](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/d0fb520acb9b9b0db58cd6d8a31ab8d7ee771bf4)

seokhong replied:
도메인과 뷰의 밸리데이션은 서로 다른 목적을 가질 수 있는데요. 관련해서 `도메인 뷰 밸리데이션` 과 같은 키워드로 학습해보시는 것도 도움이 될 것 같습니다.

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+	}
+
+	private List<CarState> findWinner() {
+		Collections.sort(cars);
```
객체를 정렬하는 과정은 O(n log n)의 시간 복잡도를 가질 것으로 보이네요.
전체 자동차 목록을 확인하여 가장 멀리 간 차의 위치를 확인한 후, 다시 전체 목록을 확인해서 해당 위치에 있는 차만 가지고 오면 O(n)의 복잡도로 문제를 해결할 수 있지 않을까요? :)

iimpala replied:
max를 활용해서 우승자를 찾는 방법도 있겠네요! 리뷰 감사합니다:)

[781e35b](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/781e35b2be4c15d5ab3f8cb4d91e5929af81bce1)

seokhong commented on src/main/java/racingcar/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,12 @@
+public class RandomNumberGenerator implements NumberGenerator {
+
+	private final Random random = new Random();
+	@Override
```
빈 줄을 추가하면 가독성이 더 좋아질 것 같아요.

iimpala replied:
반영했습니다!

[020271f](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/020271f8535a3fbebb289eb11a64288ceaf835be)

seokhong commented on src/main/java/racingcar/dto/CarState.java:
```diff
@@ -0,0 +1,20 @@
+package racingcar.dto;
+
+public class CarState {
```
1. `CarState`와 `Car`는 무슨 차이일까요? dto의 목적으로 만든 클래스라면, `CarDto`와 같은 이름을 사용해보는 건 어떨까요?
2. `Car`를 직접 사용하지 않고, dto 를 사용하도록 하신 이유가 있을까요?

iimpala replied:
저도 CarDto와 CarState 중에서 고민했었는데 Dto라는 이름을 클래스에 직접적으로 사용하는 일이 자주 있을지 몰라서 좀 더 의미를 가지는 CarState로 정했습니다.

혹시 이에 대해서 어떤 생각이신가요??

seokhong replied:
`State`, `Status`는 어떤 상태를 나타내는 클래스의 이름처럼 보이기도 합니다. 이름만 봤을 때는 '이 클래스가 자동차의 이동 상태 등을 나타내는 것일까?' 라는 생각이 먼저 떠올랐던 것 같습니다. :)
오히려 dto 라는 이름을 직접 사용하는 것이, 이 클래스의 이름을 보고 목적을 이해하는데 도움이 될 것 같다는 생각입니다!

seokhong commented on src/main/java/stringcalculator/StringParser.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.regex.Pattern;
+
+public class StringParser {
+	private static final String DEFAULT_DELIMITER = ":|,";
```
적절한 상수 사용 좋습니다. 👍

seokhong commented on src/main/java/stringcalculator/StringParser.java:
```diff
@@ -0,0 +1,49 @@
+	private static final String DEFAULT_DELIMITER = ":|,";
+	public static List<Integer> parse(String input) {
+		String delimiter = DEFAULT_DELIMITER;
+		Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(input);
```
`Pattern.compile`의 결과를 상수로 만들어서 사용하는 건 어떨까요? `parse` 메서드가 호출될 때마다 매번 패턴을 새로 만들지는 않아도 괜찮을 것 같아요!
또한 정규표현식이 나타내는 의미도 있기 때문에, 상수로 분리해서 이름을 부여하면 더 가독성 좋은 코드가 될 것 같아요.

iimpala replied:
Patter 클래스를 많이 사용해보지 않아서 예제대로 사용했는데, 관련해서 공부해봐야겠네요. 리뷰 감사합니다:)

[7573235](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/7573235ec3684d80225e2fa0bfb06a1d110a1b1c)

seokhong commented on src/main/java/stringcalculator/StringParser.java:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	private static String convertBlank(String number) {
+		if (number == null || number.isEmpty()) {
```
null 또는 값이 없을 경우 0으로 처리하도록 구현하신 이유가 있을까요? 잘못된 값이 들어왔을 때 정상적으로 동작되도록 처리하는 것과 예외 상황임을 알려주는 것 중 어떻게 처리하는 것이 좋을까요?

iimpala replied:
요구사항중에 빈 입력값은 0으로 치환하라는 내용이 있어서 정상적으로 동작하도록 구현했습니다!

seokhong commented on src/test/java/racingcar/RandomNumberGeneratorTest.java:
```diff
@@ -0,0 +1,16 @@
+	@Test
+	@DisplayName("0 ~ 9 사이의 정수를 반환한다")
+	void generateTest() {
+		NumberGenerator numberGenerator = new RandomNumberGenerator();
```
난수를 생성하는 로직에 대한 테스트를 추가하신 이유가 있을까요?

iimpala replied:
0 ~ 9범위의 난수가 생성되는지 확인하려는 것이었는데, 생각해보니 이건 Random 라이브러리 자체에 대한 테스트가 되어서 큰 의미를 가지지는 않을 것 같네요. 리뷰 감사합니다:)

[82e779e](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/82e779e8b07604ec517ec3530f92f4b53ed4b32a)

seokhong commented on src/test/java/stringcalculator/StringParserTest.java:
```diff
@@ -0,0 +1,77 @@
+public class StringParserTest {
+	@Test
+	@DisplayName("default 구분자 테스트")
+	public void parseDefaultDelimiterTest() {
```
다양한 케이스에 대한 테스트 좋습니다. 👍 
`@ParameterizedTest`를 활용해보는 건 어떨까요?

iimpala replied:
expected 파라미터로 integer값을 넘겨주는 방법을 몰라 @ParameterizedTest를 도입하려다 실패했는데 @ValueSource를 사용하면 가능할 것 같네요. 리뷰 감사합니다:)

[c32abbf](https://github.com/next-step/java-racingcar-kakao/pull/26/commits/c32abbf53e973fc6670b29059bd012d207625ced)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+	private void initRounds() {
+		boolean validationPass = false;
+		String input = "";
+		while(!validationPass){
```
while 키워드와 여는 소괄호 사이, 닫는 소괄호와 여는 중괄호 사이에 공백이 누락됐네요.
IDE를 사용하신다면 파일을 저장하는 시점에 자동으로 줄맞춤이 되도록 설정해보시거나, 줄맞춤 단축키를 사용하는 습관을 들이셔도 도움이 될 것 같아요!

iimpala replied:
IDE에 그런 기능이 있는지는 몰랐네요. 한번 찾아보겠습니다. 리뷰 감사합니다:)

seokhong reviewed (changes requested):
동현님 안녕하세요.
자동차 경주 미션을 함께 할 리뷰어 김석홍입니다. 만나서 반갑습니다!

코멘트 남긴 내용 확인 부탁드립니다.
코멘트 관련하여 궁금하신 점은 PR에 코멘트 또는 슬랙으로 DM 보내주시면 확인하는대로 회신 드리겠습니다. 🙂

iimpala committed:
style(README) todo list 형식 수정

Todo List에 checkbox 적용

iimpala committed:
fix(gradle) 해결되지 않은 confilct 수정

merge 과정에서 누락된 conflict 해결

iimpala committed:
refactor(ProceedLogic) PROCEED_THRESHOLD 상수 추출

자동차의 전진여부를 결정하는 정수를 상수로 추출하여 이름을 부여

iimpala committed:
refactor(Car) 코드 리팩토링

- Car 클래스의 비교 메소드를 compareTo 대신 comparePosition으로 변경하여 메소드의 역할을 명확히 함
- 비교 메소드에 Integer.compare() 메소드 사용

iimpala committed:
refactor(RacingGame) 코드 리팩토링

- 메소드 순서를 호출 depth 순으로 조정
- cars의 초기화를 필드에서 하도록 조정
- 스트림 메소드를 조정하여 가독성 향상
- 검증 메소드 이름 변경
- 우승자 선정 알고리즘 변경

iimpala committed:
style(RandomNumberGenerator) 멤버필드와 메소드 사이에 공백라인 추가

iimpala committed:
refactor(RacingGameRule) 전진여부를 판단하는 로직을 인터페이스화

- RacingGameRule 인터페이스를 통해 자동차의 전진여부를 판단하는 룰을 외부에서 주입할 수 있도록 설정
- 추후 요구사항이 변경되었을 때 기존코드의 수정 없이 RacingGameRule 인터페이스를 구현하여 룰 수정이 가능하도록 설계

iimpala committed:
refactor(StringCalculatorTest) 테스트코드 중복 제거

- @ParameterizedTest와 @CsvSource를 사용하여 테스트코드 내의 중복을 제거

iimpala committed:
refactor(RandomNumberGeneratorTest) 불필요한 테스트 코드 제거

- 난수생성기 테스트코드 제거

iimpala committed:
refactor(Car) 도메인 레이어에 자동차 이름 검증 로직 추가

iimpala committed:
build(gradle) java 버전 다운그레이드 및 컴파일 에러 수정

- JDK 17 -> JDK 11

iimpala committed:
refactor(StringParserTest) 테스트 코드 중복 제거

iimpala committed:
refactor(StringParser) regex 상수화

- 커스텀 구분자를 입력받을때 사용하는 regex 패턴을 상수로 추출함

iimpala commented:
수정사항 확인해보시고 다음단계 진행을 위해 머지 부탁드립니다
감사합니다:)

seokhong commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,169 @@
+
+import static racingcar.Car.MAX_CAR_NAME_LENGTH;
+
+public class RacingGame {
```
`RacingGame`이 컨트롤러의 역할과 도메인의 역할을 모두 가지고 있는 것으로 보이네요.
다음 단계에서는 `RacingGame`이 도메인의 역할에 집중하고, `RacingGameController` 와 같은 클래스를 만들어서 `RacingGame`은 `RacingGameUi`에 의존하지 않는 형태로 리팩터링 해보시면 좋겠습니다. :)

또한, 테스트도 좀 더 다양한 케이스에 대해 작성할 수 있을 것으로 보이는데요.
1. 자동차 경주의 매 라운드에 자동차에 대한 난수를 생성하고, 이동 대상에게 이동을 요청해서 이동시키는지 여부를 검증하기
2. 라운드를 진행한 후 우승자를 확인했을 때 기대한대로 우승자가 조회되는지 검증하기

이렇게 2개의 테스트는 꼭 작성해 보시면 좋겠습니다.

seokhong reviewed (approved):
동현님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
1단계 PR은 머지하겠습니다.

질문 남겨주신 부분에 대한 제 생각, 그리고 다음 단계 미션 진행하실 때 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다. 참고 부탁드립니다.
2단계 미션에서 뵙겠습니다. 🙂
